### PR TITLE
BUG: Allow numeric ExtensionDtypes in DataFrame.select_dtypes

### LINKED ISSUE #35340: ENH: Select numeric ExtensionDtypes with DataFrame.select_dtypes
#### Is your feature request related to a problem?

I would like to plot ExtensionArrays. DataFrame.select_dtypes(np.number) is called when plotting, which filters out ExtensionArrays. 

#### Describe the solution you'd like

DataFrame.select_dtypes(np.number) should select columns where ExtensionDtype._is_numeric is True, suggested here https://github.com/pandas-dev/pandas/issues/26173#issuecomment-503761441 

#### API breaking implications

Not considered.

#### Describe alternatives you've considered

Not considered.

#### Additional context

N/A

### PR DESCRIPTION
**Pull request by arw2019**
- [x] closes #35340
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Picking up #35341 (original PR addressed all comments - here I merged master + added whatsnew note)

cc @simonjayhawkins @jreback @jbrockmendel

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -3711,13 +3713,7 @@ def extract_unique_dtypes_from_dtypes_set(
...
                 unique_dtype
                 for unique_dtype in unique_dtypes
-                # error: Argument 1 to "tuple" has incompatible type
-                # "FrozenSet[Union[ExtensionDtype, str, Any, Type[str],
-                # Type[float], Type[int], Type[complex], Type[bool]]]";
-                # expected "Iterable[Union[type, Tuple[Any, ...]]]"
-                if issubclass(
-                    unique_dtype.type, tuple(dtypes_set)  # type: ignore[arg-type]
-                )
+                if np_issubclass_compat(unique_dtype, dtypes_set)
```
Comment (jreback) [inline pandas/core/frame.py:3716]: i don't think we need to create yet another method here. is there a reason you cannot just use something like

```
return issubclass(unique_dtype.type, tuple(dtypes_set)) or (
        np.number in dtypes_set and is_numeric_dtype(unique_dtype))
    )
```

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 1 above)
Comment (arw2019) [reply pandas/core/frame.py]: no reason. Done

Comment (arw2019) [reply pandas/core/frame.py]: Having now tried this I think the reason might be mypy
```
pandas/core/frame.py:3715: error: Argument 1 to "tuple" has incompatible type "FrozenSet[Union[ExtensionDtype, str, Any, Type[str], Type[float], Type[int], Type[complex], Type[bool], Type[object]]]"; expected "Iterable[Union[type, Tuple[Any, ...]]]"  [arg-type]

```

Comment (arw2019) [reply pandas/core/frame.py:3715]: Reverted to having a separate method

Comment (jreback) [reply pandas/core/frame.py:3715]: just ignore mypy here, this makes groking the code way more complicated.

Comment (arw2019) [reply pandas/core/frame.py:3715]: ok!

File: pandas/core/frame.py [hunk 2]
```diff
@@ -3711,12 +3711,15 @@ def extract_unique_dtypes_from_dtypes_set(
...
-                # expected "Iterable[Union[type, Tuple[Any, ...]]]"
-                if issubclass(
-                    unique_dtype.type, tuple(dtypes_set)  # type: ignore[arg-type]
+                if (
+                    issubclass(
+                        unique_dtype.type, tuple(dtypes_set)  # type: ignore[arg-type]
+                    )
+                    or (
+                        np.number in dtypes_set
+                        and hasattr(unique_dtype, "_is_numeric")  # is an extensionarray
```
Comment (jreback) [inline pandas/core/frame.py:3720]: can u check is_extension_array_dtype here instead

File: pandas/tests/frame/methods/test_select_dtypes.py [hunk 3]
```diff
@@ -322,3 +359,14 @@ def test_select_dtypes_typecodes(self):
         expected = df
         FLOAT_TYPES = list(np.typecodes["AllFloat"])
         tm.assert_frame_equal(df.select_dtypes(FLOAT_TYPES), expected)
+
+    def test_select_dtypes_numeric(self):
+        # GH 35340
```
Comment (jreback) [inline pandas/tests/frame/methods/test_select_dtypes.py:364]: can u add the test from the OP

i think was about int / int32

File: doc/source/whatsnew/v1.2.0.rst [hunk 4]
```diff
@@ -623,6 +623,7 @@ Numeric
 - Bug in :meth:`DataFrame.std` with ``timedelta64`` dtype and ``skipna=False`` (:issue:`37392`)
 - Bug in :meth:`DataFrame.min` and :meth:`DataFrame.max` with ``datetime64`` dtype and ``skipna=False`` (:issue:`36907`)
 - Bug in :meth:`DataFrame.idxmax` and :meth:`DataFrame.idxmin` with mixed dtypes incorrectly raising ``TypeError`` (:issue:`38195`)
+- Bug in :meth:`DataFrame.select_dtypes` with ``include=np.number`` now retains numeric ``ExtensionDtype`` columns (:issue:`35340`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:626]: move to 1.3

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/methods/test_select_dtypes.py (hunk 3 above)
Comment (arw2019) [reply pandas/tests/frame/methods/test_select_dtypes.py:364]: hmm so the OP was angling plotting EAs but it's quite old (April 2019)

With `int`/`int32`/`Int32`/`Int64` etc. these snippets pass on master, and even on 1.0.5:
```
In [21]:         ea = pd.array([1, 2], dtype=np.int32) 
    ...:         df = pd.DataFrame(ea) 
    ...:         assert df.select_dtypes(np.number).shape == df.shape                                                                              

```
but I'm not sure we test this so might add that here

Comment (arw2019) [reply pandas/tests/frame/methods/test_select_dtypes.py:364]: the ExtensionDtype derived class doesn't pass on master so I think that's what's being fixed here

File: doc/source/whatsnew/v1.2.0.rst (hunk 4 above)
Comment (arw2019) [reply doc/source/whatsnew/v1.2.0.rst:626]: Done

File: pandas/core/frame.py (hunk 2 above)
Comment (arw2019) [reply pandas/core/frame.py]: Done

File: pandas/tests/frame/methods/test_select_dtypes.py (hunk 3 above)
Comment (arw2019) [reply pandas/tests/frame/methods/test_select_dtypes.py:374]: Added those

**Comment by arw2019** [comment]
Green + addressed comments

File: pandas/core/frame.py [hunk 5]
```diff
@@ -3716,12 +3716,15 @@ def extract_unique_dtypes_from_dtypes_set(
...
-                if issubclass(
-                    unique_dtype.type, tuple(dtypes_set)  # type: ignore[arg-type]
+                if (
+                    issubclass(
+                        unique_dtype.type, tuple(dtypes_set)  # type: ignore[arg-type]
+                    )
+                    or (
+                        np.number in dtypes_set
+                        and is_extension_array_dtype(unique_dtype)
+                        and unique_dtype._is_numeric
```
Comment (jreback) [inline pandas/core/frame.py:3726]: use `getattr(unique_dtype, '_is_numeric', False)` as we don't actually require this on an EA type (maybe we should though)

File: pandas/tests/frame/methods/test_select_dtypes.py [hunk 6]
```diff
@@ -322,3 +359,24 @@ def test_select_dtypes_typecodes(self):
...
+        ),
+    )
+    def test_select_dtypes_numeric(self, arr, expected):
+        # GH 35340
+
+        df = DataFrame(arr)
+        is_selected = df.select_dtypes(np.number).shape == df.shape
+        assert is_selected == expected
+
+        # da = DummyArray([1, 2], dtype=DummyDtype(numeric=False))
```
Comment (jreback) [inline pandas/tests/frame/methods/test_select_dtypes.py:380]: remove commented code

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 5 above)
Comment (arw2019) [reply pandas/core/frame.py]: Done

File: pandas/tests/frame/methods/test_select_dtypes.py (hunk 6 above)
Comment (arw2019) [reply pandas/tests/frame/methods/test_select_dtypes.py]: Done

**Comment by arw2019** [comment]
Green + addressed comments

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @arw2019

**Pull request merged**